### PR TITLE
[Merged by Bors] - refactor(Bicategory/Functor): refactor prelaxfunctors

### PR DESCRIPTION
**Pull request by callesonne**
This PR suggests a refactor (and renaming) of prelax functors. This is to provide more reusability of code between oplax functors, pseudofunctors (and soon) lax functors.

This is done in the following way:
- A new file is created called `Bicategory.Functor.Prelax`.
- What (before this suggested refactor) was called `PrelaxFunctor`, is renamed to `LaxPrefunctor`. Recall that this is a prefunctor that is also equipped with a map `map₂` on 2-morphisms.
- A new structure is added, taking over the name `PrelaxFunctor`, which additionally requires that the associated prefunctors between hom types are functors. In other words, `map₂` should respect identities and compositions.
- Both `Pseudofunctor` and `OplaxFunctor` are updated to extend the new notion of `PrelaxFunctor`.

Finally:
Coercions have also been removed between these various structures. This is to prevent the following behaviour in the current library: given a pseudofunctor `F` the infoview will print `F.map` as `(↑F.toPrelaxfunctor).map` which makes long expressions unecessarily hard to read. The change is just like how `Functor` does not have a coercion to `PreFunctor`.

This refactor will allow us to define `mapFunctor` and `map₂Iso` on the level of `PrelaxFunctor`s, which allows it to be reused for both `Pseudofunctor`s and `OplaxFunctor`s instead of having to be defined separately.

A new constructors on this level has also been added, `PrelaxFunctor.mkOfHomFunctors`, that constructs a `PrelaxFunctor` given a map on types and functors between the corresponding hom types. This could for example be used to construct the pseudofunctor in the bicategorical yoneda lemma, by using the collection of functors `precomposing` (see `Bicategory.Basic`).

The file `CategoryTheory/WithTerminal` has been edited just to fix the proof after these new changes.

---

- [x] depends on: #13900

I am not fussed about the specific names I've chosen, and I am very happy to hear other suggestions. Also, I was not there for the mathport, so I am unsure exactly what to do with many of the porting notes and `#align` now that `PrelaxFunctor` has changed. I have removed some of them that didn't feel relevant were anymore, but I'm not sure if I was actually supposed to do that.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by yuma-mizuno — COMMENTED** [review]
Thanks for doing this! This is definitely something I left unfinished.

I feel `LaxPreFunctor/PrelaxFunctor` might have better names as it's hard to see which is the parent. How about `PrelaxFunctorStruct/PrelaxFunctor` or `PrePrelaxFunctor/PrelaxFunctor`?

**Comment by callesonne** [comment]
> Thanks for doing this! This is definitely something I left unfinished.
> 
> I feel `LaxPreFunctor/PrelaxFunctor` might have better names as it's hard to see which is the parent. How about `PrelaxFunctorStruct/PrelaxFunctor` or `PrePrelaxFunctor/PrelaxFunctor`?

Thanks for writing this library in the first place! It has been very helpful for the development of fibered categories :)

File: Mathlib/CategoryTheory/WithTerminal.lean [hunk 1]
```diff
@@ -189,51 +188,62 @@ def pseudofunctor : Pseudofunctor Cat Cat where
...
+        prelaxfunctor_toLaxPreFunctor_map₂, map₂_app, Cat.whiskerLeft_app, mapComp_hom_app,
+        Iso.refl_hom, mapComp_inv_app, Iso.refl_inv, Category.comp_id, Category.id_comp]
     · rfl
   map₂_whisker_right := by
     intros
     apply NatTrans.ext
     funext X
     cases X
-    · rw [NatTrans.comp_app, NatTrans.comp_app]
+    · dsimp
```
[the remark this answers is not in the document]
Comment (callesonne) [reply Mathlib/CategoryTheory/WithTerminal.lean]: Fixed!

**Review by yuma-mizuno — APPROVED** [review]

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**